types/mempool/priority.go

```diff
@@ -0,0 +1,234 @@
+func WithOnRead(onRead func(tx Tx)) PriorityMempoolOption {
+	return func(mp *priorityMempool) {
+		mp.onRead = onRead
+	}
+}
+
+// DefaultPriorityMempool returns a priorityMempool with no options.
+func DefaultPriorityMempool() Mempool {
+	return NewPriorityMempool()
+}
+
+// NewPriorityMempool returns the SDK's default mempool implementation which returns txs in a partial order
+// by 2 dimensions; priority, and sender-nonce.
+func NewPriorityMempool(opts ...PriorityMempoolOption) Mempool {
+	mp := &priorityMempool{
+		priorityIndex: huandu.New(huandu.LessThanFunc(txKeyLess)),
+		senderIndices: make(map[string]*huandu.SkipList),
+		scores:        make(map[txKey]int64),
+	}
+	for _, opt := range opts {
+		opt(mp)
+	}
+	return mp
+}
+
+// Insert attempts to insert a Tx into the app-side mempool in O(log n) time, returning an error if unsuccessful.
+// Sender and nonce are derived from the transaction's first signature.
+// Transactions are unique by sender and nonce.
+// Inserting a duplicate tx is an O(log n) no-op.
+// Inserting a duplicate tx with a different priority overwrites the existing tx, changing the total order of
+// the mempool.
+func (mp *priorityMempool) Insert(ctx sdk.Context, tx Tx) error {
+	sigs, err := tx.(signing.SigVerifiableTx).GetSignaturesV2()
+	if err != nil {
+		return err
+	}
+	if len(sigs) == 0 {
+		return fmt.Errorf("tx must have at least one signer")
+	}
+
+	sig := sigs[0]
+	sender := sig.PubKey.Address().String()
+	nonce := sig.Sequence
+	tk := txKey{nonce: nonce, priority: ctx.Priority(), sender: sender}
+
+	senderIndex, ok := mp.senderIndices[sender]
+	if !ok {
+		senderIndex = huandu.New(huandu.LessThanFunc(func(a, b any) int {
+			return huandu.Uint64.Compare(b.(txKey).nonce, a.(txKey).nonce)
+		}))
+		// initialize sender index if not found
+		mp.senderIndices[sender] = senderIndex
+	}
+
+	// Since senderIndex is scored by nonce, a changed priority will overwrite the existing txKey.
+	senderIndex.Set(tk, tx)
+
+	// Since mp.priorityIndex is scored by priority, then sender, then nonce, a changed priority will create a new key,
+	// so we must remove the old key and re-insert it to avoid having the same tx with different priorityIndex indexed
+	// twice in the mempool.  This O(log n) remove operation is rare and only happens when a tx's priority changes.
+	sk := txKey{nonce: nonce, sender: sender}
+	if oldScore, txExists := mp.scores[sk]; txExists {
+		mp.priorityIndex.Remove(txKey{nonce: nonce, priority: oldScore, sender: sender})
+	}
+	mp.scores[sk] = ctx.Priority()
+	mp.priorityIndex.Set(tk, tx)
+
+	return nil
+}
+
+// Select returns a set of transactions from the mempool, ordered by priority and sender-nonce in O(n) time.
+// The passed in list of transactions are ignored.  This is a readonly operation, the mempool is not modified.
+// maxBytes is the maximum number of bytes of transactions to return.
+func (mp *priorityMempool) Select(_ [][]byte, maxBytes int64) ([]Tx, error) {
```

Comment:
In general I still find this implementation somewhat confusing...specifically the for loop over `senderNode`.

At least in my head, the way I imagined this flow would look like is as follows (pseudo-code):

```
for !priorityIndex.Empty() || maxBytesReached {
  tx <- selectHighestPriorityTx()
	senderIndex <- selectSenderNode(tx.sender)

	if tx.Priority == SenderIndex.Priority {
		append(tx)
		maxBytes += tx.Size()
		moveSenderCursor(tx.sender)
	} else {
		// The highest priority tx would conflict with the sender's required nonce,
		// so we need to move onto the next highest priority transaction.
	}
}
```

Is this essentially what the existing `Select` implementation does or is my approach not correct? If they are the same approach, more or less, then I would argue the above is easier to reason about.

Reply:
I would need to see the implementations of both `selectHighestPriorityTx` and `selectSenderNode` to let you know if they're the same, they are important details.  

We *cannot* drive iteration by tx priority alone since within a given sender nonce takes precedence over priority, which is why there is a sender sub-tree for each sender and we iterate over it until the traversal is invalid due to a higher priority tx in another sender sub-tree.

Reply:
> We cannot drive iteration by tx priority alone since within a given sender nonce takes precedence over priority

You can. The general idea is "select the highest priority tx and ensure the nonce is also correct, if not, move onto next priority tx." 

You don't need to ever iterate of the sender index (apart from moving the cursor).

Reply:
Remember that there could be *multiple* same sender txs satisfying the current priority cursor which should be processed in nonce order, hence the need to iterate over a sender index.

Reply:
> select the highest priority tx and ensure the nonce is also correct, if not, move onto next priority tx. Repeat this process until exhaustion.

> > ... Iteration over sender trees ensures the second part of your statement, namely

You don't need to iterate over senders at all. You just keep selecting the next highest priority index.

I will whip up something today in a temp branch...I could be wrong, but my intuition tells me it should work.

Reply:
I think how you're envisioning this the interaction between `moveSenderCursor(tx.sender)` and `selectSenderNode(tx.sender)` results in a senderIndex iteration anyway in the case of multiple low priority txs behind a higher one by nonce.

Reply:
So after doing some napkin approaches, in order to go with a simple single loop-based approach, you'll have to temporarily "mark" the priority indices and move the cursor to the beginning. This will work, however, it will also require additional tweaking to handle the case that @amaurym pointed out.

I don't want to block on this, so you should tweak the existing impl to account for weights/smart-picking, but I will still work on a side approach to demonstrate an alternative that I think it easier to comprehend.

Reply:
@kocubinski no, there is still never any iteration of sender indexes. Just moving the cursor and marking them.

Reply:
@alexanderbez Isn't iteration just moving a cursor?

Reply:
yes, but the implementation I think is simpler as you just move two cursors. I find your inner loop over sender cursor to be pretty confusing personally. But it's not a deal breaker.